aten/src/ATen/Utils.h

```diff
@@ -93,10 +93,21 @@ inline int64_t sum_intlist(ArrayRef<int64_t> list) {
   return std::accumulate(list.begin(), list.end(), 0ll);
 }
 
-inline int64_t prod_intlist(ArrayRef<int64_t> list) {
-  return std::accumulate(list.begin(), list.end(), 1ll, std::multiplies<int64_t>());
+// inline int64_t prod_intlist(ArrayRef<int64_t> list) {
+//   return std::accumulate(list.begin(), list.end(), 1ll, std::multiplies<int64_t>());
+// }
+
+template<typename C>
+inline typename C::value_type prod_intlist(const C &container){
+    typedef typename C::value_type element_type;
+    return std::accumulate(container.begin(), container.end(), static_cast<element_type>(1), std::multiplies<element_type>());
```

Comment:
My worry here is that you can have a vector with `int` several times where multiplying those together gives a value too large for `int`, so it might be better to use `int64_t` directly?

Reply:
The function is templated by the return type so if user is worried about overflowing they should store the result in an int64_t type. My worry is what would the type be in something like:

```
int i = 1;
int64_t j = i * prod_intlist(...)
```

Would the result of prod_intlist be int or int64_t?

Reply:
`i * prod_intlist(...)` returns `int64_t` per [integer promotion](https://en.cppreference.com/w/c/language/conversion).

However, it doesn't matter what the return type is since `std::accumulate` templates to the type of the initial element (`static_cast<element_type>(1)`) here. If that's too small to hold the accumulated value then `std::accumulate` returns the wrong value and therefore `prod_intlist` does as well.